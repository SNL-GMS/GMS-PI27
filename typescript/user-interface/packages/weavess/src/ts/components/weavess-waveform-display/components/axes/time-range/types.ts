import type { WeavessTypes } from '@gms/weavess-core';

export interface TimeRangeProps {
  labelWidthPx?: number;
  currentZoomInterval: WeavessTypes.TimeRange;
}
