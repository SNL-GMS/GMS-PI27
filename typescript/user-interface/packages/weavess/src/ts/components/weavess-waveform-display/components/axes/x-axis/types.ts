import type { WeavessTypes } from '@gms/weavess-core';

export interface XAxisProps {
  /** waveform interval loaded and available to display */
  displayInterval: WeavessTypes.TimeRange;

  /** Add border to top */
  borderTop: boolean;

  /** Label width in px */
  labelWidthPx: number;

  /** Scrollbar width in px */
  scrollbarWidthPx: number;

  /** (optional) x-axis label  */
  label?: string;
}

export interface XAxisState {}
