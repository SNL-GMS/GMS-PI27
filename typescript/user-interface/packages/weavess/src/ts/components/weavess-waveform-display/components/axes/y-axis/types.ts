export interface YAxisProps {
  /** The height in percentage */
  heightInPercentage: number | undefined;

  /** Max amplitude as a number */
  maxAmplitude: number | undefined;

  /** Min amplitude as a number */
  minAmplitude: number | undefined;

  /** the y axis ticks */
  yAxisTicks?: number[];
}

export interface YAxisState {}
