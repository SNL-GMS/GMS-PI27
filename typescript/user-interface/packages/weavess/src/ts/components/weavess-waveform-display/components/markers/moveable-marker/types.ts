import type { WeavessTypes } from '@gms/weavess-core';

export interface MoveableMarkerProps {
  /** The marker config */
  marker: WeavessTypes.Marker;

  /** The associated start marker (usually only for a selection window) */
  associatedStartMarker?: WeavessTypes.Marker;

  /** The associated end marker (usually only for a selection window) */
  associatedEndMarker?: WeavessTypes.Marker;

  /** Percentage Location 0-100 as number */
  percentageLocation: number;

  /** Label Width in px */
  labelWidthPx: number;

  /** React key but called name since key was causing fortify issues */
  name?: string;

  /**
   * Minimum amount of time in seconds between this and a possible corresponding marker
   * (e.g.- start and end marker for selection windows) and callback to trigger when reached
   */
  minimumSelectionWindowDuration?: { durationInSeconds: number; onDurationReached: () => void };

  /** Start and end of the entire time range */
  timeRange(): WeavessTypes.TimeRange;

  /** The ratio that the waveform panel is zoomed in: the zoom interval divided
   * by the total viewable interval. This is a unitless value. */
  getZoomRatio(): number;

  /** Returns container client width */
  containerClientWidth(): number;

  /** Returns the viewPort client width */
  viewportClientWidth(): number;

  /**
   * (optional) updates the location of the marker
   *
   * @param marker the marker
   */
  onUpdateMarker?(marker: WeavessTypes.Marker): void;

  /**  */
  updateTimeWindowSelection?(): void;
}

export interface MoveableMarkerState {
  marker: WeavessTypes.Marker;
}
