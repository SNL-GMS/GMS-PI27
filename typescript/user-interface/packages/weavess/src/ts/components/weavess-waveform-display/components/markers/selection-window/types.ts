import type { WeavessTypes } from '@gms/weavess-core';

export interface SelectionWindowProps {
  /** Selection window props as a StationWindowConfig */
  selectionWindow: WeavessTypes.SelectionWindow;

  /** Label Width in pixels */
  labelWidthPx: number;

  /** Start and end of the entire time range */
  timeRange(): WeavessTypes.TimeRange;

  /** The ratio of the zoom interval divided by the total viewable interval. Unitless. */
  getZoomRatio: () => number;

  /** Ref to the html canvas element */
  canvasRef(): HTMLCanvasElement | null;

  /** Returns container client width  */
  containerClientWidth(): number;

  /** Returns the viewPort client width */
  viewportClientWidth(): number;

  /**
   * Minimum amount of time in seconds between this and a possible corresponding marker
   * (e.g.- start and end marker for selection windows) and callback to trigger when reached
   */
  minimumSelectionWindowDuration?: { durationInSeconds: number; onDurationReached: () => void };

  /**
   * Computes the time in seconds for the mouse x position.
   *
   * @param mouseXPosition the mouse x position to compute the time on
   *
   * @returns The computed time in seconds
   */
  computeTimeSecsForMouseXPosition(mouseXPosition: number): number;

  /**
   * (optional) Event handler for invoked while the selection is moving
   *
   * @param selection the selection
   */
  onMoveSelectionWindow?(selection: WeavessTypes.SelectionWindow): void;

  /**
   * (optional) updates the location of the marker
   *
   * @param selection the selection
   */
  onUpdateSelectionWindow?(selection: WeavessTypes.SelectionWindow): void;

  /**
   * (optional) Event handler for click events within a selection
   *
   * @param selection the selection
   * @param timeSecs epoch seconds of where drag ended in respect to the data
   */
  onClickSelectionWindow?(selection: WeavessTypes.SelectionWindow, timeSecs: number): void;

  /**
   * Mouse Move event
   *
   * @param e mouse event as React.MouseEvent<HTMLDivElement>
   */
  onMouseMove(e: React.MouseEvent<HTMLDivElement>): void;

  /**
   * Mouse down event
   *
   * @param e React.MouseEvent<HTMLDivElement>
   */
  onMouseDown(e: React.MouseEvent<HTMLDivElement>): void;

  /**
   * Mouse up event
   *
   * @param e React.MouseEvent<HTMLDivElement>
   */
  onMouseUp(e: React.MouseEvent<HTMLDivElement>): void;
}

export interface SelectionWindowState {
  selectionWindow: WeavessTypes.SelectionWindow;
}

export enum SelectionWindowBoundaryMarkerType {
  LEAD,
  LAG
}

export interface SelectionWindowBoundaryMarkerProps {
  // /** type of marker to determine which side of the selection window we are dragging */
  boundaryType: SelectionWindowBoundaryMarkerType;

  /** epoch time in seconds */
  timeSecs: number;

  /** react key but called name since key was causing fortify issues */
  name?: string;

  /** Start and end of the entire time range */
  timeRange(): WeavessTypes.TimeRange;

  /**
   * (optional) updates the location of the marker
   *
   * @param marker the marker
   */
  onMarkerUpdated(boundaryType: SelectionWindowBoundaryMarkerType, timeSecs: number): void;

  /**
   * gets the time from the mouse position
   * @param clientX
   */
  getTimeSecsForClientX(clientX: number): number | undefined;

  /**
   * sets the updating property
   * @param isUpdating
   */
  setIsUpdating(isUpdating: boolean): void;
}

export interface SelectionWindowBoundaryMarkerState {
  marker: WeavessTypes.Marker;
}
