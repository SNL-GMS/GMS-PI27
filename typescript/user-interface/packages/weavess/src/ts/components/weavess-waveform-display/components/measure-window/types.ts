import type { WeavessTypes } from '@gms/weavess-core';

import type { WeavessWaveformPanelProps } from '../../types';
import type { WeavessWaveformPanel } from '../../weavess-waveform-panel';

export interface MeasureWindowProps extends WeavessWaveformPanelProps {
  /**
   * The information about the selected measure window
   */
  measureWindowSelection: WeavessTypes.MeasureWindowSelection | undefined;

  /**
   * A function that is passed the underlying WeavessWaveformPanel ref to expose the
   * WeavessWaveformPanel's public functions and DOM info
   */
  setMeasureWindowRef: (ref: WeavessWaveformPanel | null) => void;

  /**
   * The height of the measure window, including the x-axis.
   */
  measureWindowHeightPx: number;
}
