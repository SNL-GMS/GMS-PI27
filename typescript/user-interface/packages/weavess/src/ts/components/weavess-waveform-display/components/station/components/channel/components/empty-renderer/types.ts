import type { WeavessTypes } from '@gms/weavess-core';

export interface EmptyRendererProps {
  /** waveform interval loaded and available to display */
  displayInterval: WeavessTypes.TimeRange;
}

export interface EmptyRendererState {}
