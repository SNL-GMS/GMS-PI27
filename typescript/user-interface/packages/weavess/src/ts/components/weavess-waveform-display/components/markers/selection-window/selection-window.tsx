/* eslint-disable react/destructuring-assignment */
import { UILogger } from '@gms/ui-util';
import type { WeavessTypes } from '@gms/weavess-core';
import { WeavessConstants } from '@gms/weavess-core';
import debounce from 'lodash/debounce';
import isEqual from 'lodash/isEqual';
import React from 'react';

import { SingleDoubleClickEvent } from '../../../events/single-double-click-event';
import { calculateLeftPercent } from '../../../utils';
import { VerticalMarker } from '../vertical-marker';
import { SelectionWindowBoundaryMarker } from './selection-window-boundary-marker';
import type { SelectionWindowProps, SelectionWindowState } from './types';
import { SelectionWindowBoundaryMarkerType } from './types';

const logger = UILogger.create('GMS_LOG_WEAVESS', process.env.GMS_LOG_WEAVESS);

/**
 * SelectionWindow Component. Contains two moveable boundary markers.
 */
export class SelectionWindow extends React.PureComponent<
  SelectionWindowProps,
  SelectionWindowState
> {
  /** Amount of mouse movement before updating */
  private readonly MOUSE_MOVE_CONSTRAINT_SECS = 0.005; // represents 5 ms

  /** indicates if the mouse is down */
  private mouseDown = false;

  /** indicates if the mouse is dragging */
  private isUpdating = false;

  /** handler for handling single and double click events */
  private readonly handleSingleDoubleClick: SingleDoubleClickEvent = new SingleDoubleClickEvent();

  /** The number of milliseconds to delay calls to onMoveSelectionClick  */
  private readonly debouncedOnMoveSelectionClickMS: number = 500;

  /** The debounced function of onMoveSelectionClick event handler */
  private debouncedOnMoveSelectionClick:
    | ((() => void) & { cancel(): void; flush(): void })
    | undefined;

  /**
   * Constructor
   *
   * @param props Selection Window props as SelectionWindowProps
   */
  public constructor(props: SelectionWindowProps) {
    super(props);
    this.state = {
      selectionWindow: props.selectionWindow
    };
  }

  // ******************************************
  // BEGIN REACT COMPONENT LIFECYCLE METHODS
  // ******************************************

  /**
   * Called immediately after updating occurs. Not called for the initial render.
   *
   * @param prevProps the previous props
   */
  public componentDidUpdate(prevProps: SelectionWindowProps): void {
    // When not dragging update state to equal props selection window
    // constrain to within timeRange
    if (!this.isUpdating && !isEqual(this.state.selectionWindow, this.props.selectionWindow)) {
      this.setState({ selectionWindow: this.props.selectionWindow });
    }

    // eslint-disable-next-line @typescript-eslint/unbound-method
    if (!isEqual(this.props.onMoveSelectionWindow, prevProps.onMoveSelectionWindow)) {
      if (this.debouncedOnMoveSelectionClick) {
        this.debouncedOnMoveSelectionClick.cancel();
      }
      this.debouncedOnMoveSelectionClick = this.props.onMoveSelectionWindow
        ? debounce(
            () => {
              if (this.props.onMoveSelectionWindow) {
                this.props.onMoveSelectionWindow(this.getSelectionWindow());
              }
            },
            this.debouncedOnMoveSelectionClickMS,
            { maxWait: this.debouncedOnMoveSelectionClickMS }
          )
        : undefined;
    }
  }

  /**
   * Catches exceptions generated in descendant components.
   * Unhandled exceptions will cause the entire component tree to unmount.
   *
   * @param error the error that was caught
   * @param info the information about the error
   */
  public componentDidCatch(error, info): void {
    logger.error(`Weavess Selection Window Error: ${error} : ${info}`);
  }

  // ******************************************
  // END REACT COMPONENT LIFECYCLE METHODS
  // ******************************************

  // eslint-disable-next-line react/sort-comp
  public render(): JSX.Element {
    const selectionWindow = this.getSelectionWindow();

    const leftPercent = calculateLeftPercent(
      selectionWindow.startMarker.timeSecs,
      this.props.timeRange().startTimeSecs,
      this.props.timeRange().endTimeSecs
    );
    const rightPercent = calculateLeftPercent(
      selectionWindow.endMarker.timeSecs,
      this.props.timeRange().startTimeSecs,
      this.props.timeRange().endTimeSecs
    );

    return (
      // eslint-disable-next-line jsx-a11y/no-static-element-interactions
      <div
        className="selection-window"
        data-testid="selection-window"
        onMouseDown={this.onMouseDown}
        onMouseMove={this.onMouseMove}
        onMouseUp={this.onMouseUp}
      >
        {/* eslint-disable-next-line jsx-a11y/no-static-element-interactions */}
        <div
          className="selection-window-selection"
          style={{
            backgroundColor: `${selectionWindow.color}`,
            left: `${leftPercent}%`,
            right: `${WeavessConstants.PERCENT_100 - rightPercent}%`,
            cursor: selectionWindow.isMoveable ? 'move' : 'auto'
          }}
          onMouseDown={this.onSelectionWindowClick}
          onDoubleClick={this.handleSingleDoubleClick.onDoubleClick}
        />
        {this.createMarkers(selectionWindow)}
      </div>
    );
  }

  /**
   * Return the selection window depending on if we are dragging or not.
   * If not dragging then position is determined from the props
   *
   * @returns SelectionWindow
   */
  private readonly getSelectionWindow = (): WeavessTypes.SelectionWindow => {
    return this.state.selectionWindow;
  };

  /**
   * Selection window on click logic, creates mouse move and mouse down
   * Listeners to determine where to move the window and the markers.
   *
   * @param event
   */
  private readonly onSelectionWindowClick = (mouseClickEvent: React.MouseEvent<HTMLDivElement>) => {
    if (
      mouseClickEvent.button === 2 ||
      mouseClickEvent.shiftKey ||
      mouseClickEvent.ctrlKey ||
      mouseClickEvent.metaKey
    )
      return;
    const { clientX } = mouseClickEvent;
    mouseClickEvent.stopPropagation();

    if (this.mouseDown) {
      return;
    }
    this.mouseDown = true;

    // calculate initial start time
    const timeSecs = this.getTimeSecsForClientX(clientX);

    // If can't figure out where on the time line clicked
    if (!timeSecs) return;

    // Set the offset from where the mouse clicked and the start time marker of the selection window
    const startMarkerTimeSecs = this.getSelectionWindow().startMarker.timeSecs;
    const leadingEdgeClickOffset = timeSecs - startMarkerTimeSecs;

    const onMouseMove = (mouseMoveEvent: MouseEvent) => {
      // Calculate the mouse position time adjusting for start time offset
      const newStartTimeSecs =
        (this.getTimeSecsForClientX(mouseMoveEvent.clientX) ?? 0) - leadingEdgeClickOffset;
      if (this.mouseDown) {
        this.setIsUpdating(true);
        const mouseDeltaTimeSecs = newStartTimeSecs - startMarkerTimeSecs;

        // the mouse is considered to be dragging if the user has moved greater than 5ms
        if (Math.abs(mouseDeltaTimeSecs) > this.MOUSE_MOVE_CONSTRAINT_SECS) {
          this.updateSelectionWindowPosition(newStartTimeSecs);
        }
      }
    };

    const onMouseUp = (mouseUpEvent: MouseEvent) => {
      document.removeEventListener('mousemove', onMouseMove);
      document.removeEventListener('mouseup', onMouseUp);
      this.endSelectionWindowDrag(mouseUpEvent, timeSecs);
    };
    document.addEventListener('mousemove', onMouseMove);
    document.addEventListener('mouseup', onMouseUp);
  };

  /**
   * Updates the start/end marker times of the selection window in state
   * @param newStartTimeSecs amount window moved in seconds
   */
  private readonly updateSelectionWindowPosition = (newStartTimeSecs: number) => {
    const selectionWindow = this.getSelectionWindow();

    const selectionWindowDuration =
      selectionWindow.endMarker.timeSecs - selectionWindow.startMarker.timeSecs;

    const deltaTimeSecs = newStartTimeSecs - selectionWindow.startMarker.timeSecs;
    let startMarkerTime = this.constrainTimeSecs(
      newStartTimeSecs,
      selectionWindow.startMarker.minTimeSecsConstraint,
      selectionWindow.startMarker.maxTimeSecsConstraint
    );
    let endMarkerTime = this.constrainTimeSecs(
      selectionWindow.endMarker.timeSecs + deltaTimeSecs,
      selectionWindow.endMarker.minTimeSecsConstraint,
      selectionWindow.endMarker.maxTimeSecsConstraint
    );

    // window duration changed due to constraint enforcement
    if (endMarkerTime - startMarkerTime !== selectionWindowDuration) {
      // the start marker was constrained
      if (
        startMarkerTime ===
        (selectionWindow.startMarker.minTimeSecsConstraint || this.props.timeRange().startTimeSecs)
      ) {
        // reset the end marker
        endMarkerTime = startMarkerTime + selectionWindowDuration;
      } else {
        // end marker was constrained, reset the start marker
        startMarkerTime = endMarkerTime - selectionWindowDuration;
      }
    }

    this.setState(prevState => ({
      selectionWindow: {
        ...prevState.selectionWindow,
        startMarker: {
          ...prevState.selectionWindow.startMarker,
          timeSecs: startMarkerTime
        },
        endMarker: {
          ...prevState.selectionWindow.endMarker,
          timeSecs: endMarkerTime
        }
      }
    }));

    if (this.debouncedOnMoveSelectionClick) {
      this.debouncedOnMoveSelectionClick();
    }
  };

  /**
   * Handles mouse up ending drag of selection window
   *
   * @param mouseUpEvent Ending mouse drag event
   * @param endingDragTime Epoch time where drag window stops
   */
  private readonly endSelectionWindowDrag = (
    mouseUpEvent: MouseEvent,
    endingDragTime: number
  ): void => {
    if (this.debouncedOnMoveSelectionClick) {
      this.debouncedOnMoveSelectionClick.cancel();
    }

    const selectionWindow = this.getSelectionWindow();

    if (this.isUpdating && selectionWindow.isMoveable && this.props.onUpdateSelectionWindow) {
      // only update if the selection window is moveable; no false updates
      this.setIsUpdating(false);
      this.props.onUpdateSelectionWindow(selectionWindow);
    } else {
      // handle a single click event
      this.handleSingleDoubleClick.onSingleClickEvent(mouseUpEvent, () => {
        if (this.props.onClickSelectionWindow) {
          this.props.onClickSelectionWindow(selectionWindow, endingDragTime);
        }
      });
    }
    this.mouseDown = false;
  };

  /**
   * Create boarder markers
   */
  private readonly createMarkers = (
    selectionWindow: WeavessTypes.SelectionWindow
  ): JSX.Element[] => {
    if (!selectionWindow) return [];
    const boundaryMarkers: JSX.Element[] = [];
    const borderMarkersKey = 'boundary-marker-lead';
    const verticalMarkerKey = 'vertical-marker-start';
    boundaryMarkers.push(
      selectionWindow.isMoveable ? (
        <SelectionWindowBoundaryMarker
          key={selectionWindow.startMarker.id}
          name={borderMarkersKey}
          boundaryType={SelectionWindowBoundaryMarkerType.LEAD}
          timeSecs={selectionWindow.startMarker.timeSecs}
          timeRange={this.props.timeRange}
          onMarkerUpdated={this.onMarkerUpdated}
          getTimeSecsForClientX={this.getTimeSecsForClientX}
          setIsUpdating={this.setIsUpdating}
        />
      ) : (
        <VerticalMarker
          key={selectionWindow.startMarker.id}
          name={verticalMarkerKey}
          color={selectionWindow.startMarker.color}
          lineStyle={selectionWindow.startMarker.lineStyle}
          percentageLocation={calculateLeftPercent(
            selectionWindow.startMarker.timeSecs,
            this.props.timeRange().startTimeSecs,
            this.props.timeRange().endTimeSecs
          )}
        />
      )
    );
    const moveableMarkerKey = 'boundary-marker-lag';
    const verticalMarkerEndKey = 'vertical-marker-end';
    boundaryMarkers.push(
      selectionWindow.isMoveable ? (
        <SelectionWindowBoundaryMarker
          key={selectionWindow.endMarker.id}
          name={moveableMarkerKey}
          boundaryType={SelectionWindowBoundaryMarkerType.LAG}
          timeSecs={selectionWindow.endMarker.timeSecs}
          timeRange={this.props.timeRange}
          onMarkerUpdated={this.onMarkerUpdated}
          getTimeSecsForClientX={this.getTimeSecsForClientX}
          setIsUpdating={this.setIsUpdating}
        />
      ) : (
        <VerticalMarker
          key={selectionWindow.endMarker.id}
          name={verticalMarkerEndKey}
          color={selectionWindow.startMarker.color}
          lineStyle={selectionWindow.startMarker.lineStyle}
          percentageLocation={calculateLeftPercent(
            selectionWindow.endMarker.timeSecs,
            this.props.timeRange().startTimeSecs,
            this.props.timeRange().endTimeSecs
          )}
        />
      )
    );
    return boundaryMarkers;
  };

  /**
   * Handles the on update marker event and updates the selection
   */
  private readonly onMarkerUpdated = (
    boundaryType: SelectionWindowBoundaryMarkerType,
    timeSecs: number
  ) => {
    // only update if the selection window is moveable; no false updates
    const selectionWindow = this.getSelectionWindow();

    // determined which marker moved
    const startMarkerDidMove: boolean = boundaryType === SelectionWindowBoundaryMarkerType.LEAD;
    const endMarkerDidMove: boolean = boundaryType === SelectionWindowBoundaryMarkerType.LAG;

    // set to new time if moved, otherwise set from state
    let startMarkerTime = startMarkerDidMove
      ? this.constrainTimeSecs(
          timeSecs,
          this.props.selectionWindow.startMarker.minTimeSecsConstraint,
          this.props.selectionWindow.startMarker.maxTimeSecsConstraint
        )
      : selectionWindow.startMarker.timeSecs;
    let endMarkerTime = endMarkerDidMove
      ? this.constrainTimeSecs(
          timeSecs,
          this.props.selectionWindow.endMarker.minTimeSecsConstraint,
          this.props.selectionWindow.endMarker.maxTimeSecsConstraint
        )
      : selectionWindow.endMarker.timeSecs;

    if (
      this.props.minimumSelectionWindowDuration &&
      endMarkerTime - startMarkerTime < this.props.minimumSelectionWindowDuration.durationInSeconds
    ) {
      if (startMarkerDidMove) {
        startMarkerTime =
          endMarkerTime - this.props.minimumSelectionWindowDuration.durationInSeconds;
      } else {
        endMarkerTime =
          startMarkerTime + this.props.minimumSelectionWindowDuration.durationInSeconds;
      }
      this.props.minimumSelectionWindowDuration.onDurationReached();
    }
    this.setState(
      prevState => ({
        selectionWindow: {
          ...prevState.selectionWindow,
          startMarker: {
            ...prevState.selectionWindow.startMarker,
            timeSecs: startMarkerTime
          },
          endMarker: {
            ...prevState.selectionWindow.endMarker,
            timeSecs: endMarkerTime
          }
        }
      }),
      () => {
        if (!this.isUpdating && selectionWindow.isMoveable && this.props.onUpdateSelectionWindow) {
          this.props.onUpdateSelectionWindow(this.state.selectionWindow);
        }
      }
    );
  };

  /**
   * Returns the time in seconds for the given clientX.
   *
   * @param clientX The clientX
   *
   * @returns The time in seconds; undefined if clientX is
   * out of the channel's bounds on screen.
   */
  private readonly getTimeSecsForClientX = (clientX: number): number | undefined => {
    const canvasRef = this.props.canvasRef();
    if (!canvasRef) return undefined;

    const offset = canvasRef.getBoundingClientRect();
    if (clientX < offset.left && clientX > offset.right) return undefined;

    // position in [0,1] in the current channel bounds.
    const position = (clientX - offset.left) / offset.width;
    // Return computed time
    return this.props.computeTimeSecsForMouseXPosition(position);
  };

  /**
   * onMouseDown event handler.
   *
   * @param e mouse event as React.MouseEvent<HTMLDivElement>
   */
  private readonly onMouseDown = (e: React.MouseEvent<HTMLDivElement>) => {
    this.props.onMouseDown(e);
  };

  /**
   * onMouseMove event handler.
   *
   * @param e mouse event as React.MouseEvent<HTMLDivElement>
   */
  private readonly onMouseMove = (e: React.MouseEvent<HTMLDivElement>) => {
    this.props.onMouseMove(e);
  };

  /**
   * onMouseUp event handler.
   *
   * @param e mouse event as React.MouseEvent<HTMLDivElement>
   */
  private readonly onMouseUp = (e: React.MouseEvent<HTMLDivElement>) => {
    this.props.onMouseUp(e);
  };

  private readonly setIsUpdating = (isUpdating: boolean) => {
    this.isUpdating = isUpdating;
  };

  private readonly constrainTimeSecs = (
    timeSecs: number,
    minConstraint = this.props.timeRange().startTimeSecs,
    maxConstraint = this.props.timeRange().endTimeSecs
  ): number => {
    if (timeSecs < minConstraint) return minConstraint;
    if (timeSecs > maxConstraint) return maxConstraint;
    return timeSecs;
  };
}
