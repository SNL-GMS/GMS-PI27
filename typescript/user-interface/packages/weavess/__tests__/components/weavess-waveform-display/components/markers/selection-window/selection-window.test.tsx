/* eslint-disable @typescript-eslint/no-magic-numbers */
import { WeavessTypes } from '@gms/weavess-core';
import { fireEvent, render } from '@testing-library/react';
import * as React from 'react';

import { SelectionWindow } from '../../../../../../src/ts/components/weavess-waveform-display/components/markers/selection-window/selection-window';
import type { SelectionWindowProps } from '../../../../../../src/ts/components/weavess-waveform-display/components/markers/selection-window/types';
import { SelectionWindowBoundaryMarkerType } from '../../../../../../src/ts/components/weavess-waveform-display/components/markers/selection-window/types';

jest.mock('lodash/debounce', () => {
  return fn => {
    fn();
    // eslint-disable-next-line no-param-reassign
    fn.cancel = jest.fn();
    return fn;
  };
});
const selectionWindowRef = {
  containerRef: {
    style: { left: 1, right: 2 }
  },
  offsetLeft: 10
};
const testRect = {
  width: 1000,
  left: 0
};
const onMouseMove = jest.fn();
const onMouseDown = jest.fn();
const onMouseUp = jest.fn();
let position = 0;
const props: SelectionWindowProps = {
  selectionWindow: {
    id: 'selection-window',
    startMarker: {
      id: 'my-start-marker',
      color: '#ff0000',
      lineStyle: WeavessTypes.LineStyle.DASHED,
      timeSecs: 200,
      minTimeSecsConstraint: 15,
      maxTimeSecsConstraint: 600
    },
    endMarker: {
      id: 'my-end-marker',
      color: '#ff0000',
      lineStyle: WeavessTypes.LineStyle.DASHED,
      timeSecs: 400,
      minTimeSecsConstraint: 15,
      maxTimeSecsConstraint: 600
    },
    color: '#ff0000',
    isMoveable: true
  },
  minimumSelectionWindowDuration: { durationInSeconds: 1, onDurationReached: jest.fn() },
  labelWidthPx: 5,
  timeRange: jest.fn(() => ({ startTimeSecs: 0, endTimeSecs: 1000 })),
  getZoomRatio: jest.fn(() => 0.5),
  canvasRef: jest.fn().mockReturnValue({
    rect: testRect,
    getBoundingClientRect: jest.fn(() => testRect),
    offsetWidth: 1,
    offsetHeight: 1
  }),
  containerClientWidth: jest.fn(() => 800),
  viewportClientWidth: jest.fn(() => 800),
  computeTimeSecsForMouseXPosition: jest.fn(() => {
    position += 10;
    return position;
  }),
  onMoveSelectionWindow: jest.fn(),
  onUpdateSelectionWindow: jest.fn(),
  onClickSelectionWindow: jest.fn(),
  onMouseMove,
  onMouseDown,
  onMouseUp
};

const event: any = {
  clientX: 50,
  stopPropagation: jest.fn(),
  target: {
    offsetLeft: 5,
    style: { right: 1, left: 2 }
  },
  nativeEvent: {
    offsetX: 200,
    offsetY: 180
  }
};

describe('Weavess SelectionWindow Marker', () => {
  it('Weavess Selection Window to be defined', () => {
    expect(SelectionWindow).toBeDefined();
  });

  it('renders', () => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const { container } = render(<SelectionWindow {...props} />);
    expect(container).toMatchSnapshot();
  });

  it('onSelectionWindowClick', () => {
    const sw: any = new SelectionWindow(props);
    const spy = jest.spyOn(sw, 'onSelectionWindowClick');
    const spy2 = jest.spyOn(sw, 'getTimeSecsForClientX');
    sw.onSelectionWindowClick(event);
    expect(spy).toHaveBeenCalled();
    expect(spy2).toHaveBeenCalled();
    event.clientX = 500;
    const moveEvent = {
      ...event,
      clientX: 500
    };
    let mouse = new MouseEvent('mousemove', moveEvent);
    expect(document.body.dispatchEvent(mouse)).toBeTruthy();
    mouse = new MouseEvent('mouseup', moveEvent);
    expect(document.body.dispatchEvent(mouse)).toBeTruthy();
  });

  it('onSelectionWindowClick handleSingleDoubleClick', () => {
    const sw: any = new SelectionWindow(props);
    sw.onSelectionWindowClick(event);
    const mouse = new MouseEvent('mouseup', event);
    expect(document.body.dispatchEvent(mouse)).toBeTruthy();
  });

  it('onMarkerUpdated', () => {
    const sw: any = new SelectionWindow(props);
    const spy = jest.spyOn(sw, 'onMarkerUpdated');
    sw.onMarkerUpdated(SelectionWindowBoundaryMarkerType.LEAD, 1);
    expect(spy).toHaveBeenCalled();
  });

  it('updateSelectionWindowPosition', () => {
    const sw: any = new SelectionWindow(props);
    sw.selectionWindowRef = selectionWindowRef;
    sw.mouseDown = true;
    sw.onSelectionWindowClick(event);
    const spy = jest.spyOn(sw, 'updateSelectionWindowPosition');
    sw.updateSelectionWindowPosition(10);
    expect(spy).toHaveBeenCalled();
  });

  it('getTimeSecsForClientX', () => {
    const sw: any = new SelectionWindow(props);
    const spy = jest.spyOn(sw, 'getTimeSecsForClientX');
    sw.getTimeSecsForClientX(event);
    expect(spy).toHaveBeenCalled();
  });

  it('can handle drag events', () => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const result = render(<SelectionWindow {...props} />);
    const selection = result.getByTestId('selection-window');

    fireEvent.mouseDown(selection);
    fireEvent.mouseMove(selection, { clientX: 200 });
    fireEvent.mouseUp(selection);
    expect((onMouseDown.mock.calls[0][0] as React.SyntheticEvent).type).toMatchInlineSnapshot(
      `"mousedown"`
    );
    expect((onMouseMove.mock.calls[0][0] as React.SyntheticEvent).type).toMatchInlineSnapshot(
      `"mousemove"`
    );
    expect((onMouseUp.mock.calls[0][0] as React.SyntheticEvent).type).toMatchInlineSnapshot(
      `"mouseup"`
    );
  });

  it('can update onSelectionWindow function', () => {
    // eslint-disable-next-line react/jsx-props-no-spreading
    const result = render(<SelectionWindow {...props} />);
    const newOnMoveSelectionWindow = jest.fn();
    result.rerender(
      // eslint-disable-next-line react/jsx-props-no-spreading
      <SelectionWindow {...{ ...props, onMoveSelectionWindow: newOnMoveSelectionWindow }} />
    );
    expect(newOnMoveSelectionWindow).toHaveBeenCalledTimes(1);
  });

  it('componentDidCatch', () => {
    const instance = new SelectionWindow(props);
    const spy = jest.spyOn(instance, 'componentDidCatch');
    instance.componentDidCatch(new Error('error'), { componentStack: undefined });
    expect(spy).toHaveBeenCalled();
  });
});
