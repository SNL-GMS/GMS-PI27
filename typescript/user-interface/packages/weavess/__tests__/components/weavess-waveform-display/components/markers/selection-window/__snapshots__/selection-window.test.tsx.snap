// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Weavess SelectionWindow Marker renders 1`] = `
<div>
  <div
    class="selection-window"
    data-testid="selection-window"
  >
    <div
      class="selection-window-selection"
      style="background-color: rgb(255, 0, 0); left: 20%; right: 60%; cursor: move;"
    />
    <div
      class="moveable-marker"
      data-testid="selection-window-boundary-marker"
      style="left: 20%;"
    />
    <div
      class="moveable-marker"
      data-testid="selection-window-boundary-marker"
      style="left: 40%;"
    />
  </div>
</div>
`;
